micrometer-observation-datasource/src/main/java/io/micronaut/micrometer/observation/datasource/ProxyDataSourceResolver.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2017-2025 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.micrometer.observation.datasource;
+
+import io.micronaut.context.annotation.Primary;
+import io.micronaut.jdbc.DataSourceResolver;
+import jakarta.inject.Singleton;
+import net.ttddyy.dsproxy.support.ProxyDataSource;
+
+import javax.sql.DataSource;
+import java.util.List;
+
+/**
+ * A custom implementation of {@link DataSourceResolver} that resolves the underlying target data source
+ * even when it has been wrapped by proxying or instrumentation logic.
+ * This resolver first delegates to another {@link DataSourceResolver} instance and then checks if the
+ * resolved data source is an instance of {@link ProxyDataSource}. If so, it returns the actual data
+ * source wrapped by the proxy.
+ *
+ * @author radovanradic
+ * @since 5.10
+ */
+@Singleton
+@Primary
+public class ProxyDataSourceResolver implements DataSourceResolver {
```

Comment:
should this be here or upstream in `Micronaut-sql`?

Reply:
should it be `public`? Is it public API?

Reply:
Removed `public`, not needed. I think it should not go to micronaut-sql because it would introduce dependency on datasource-micrometer and now will be activated only if this module is added so I would leave it here.

Something similar might be needed for micronaut-tracing `OpenTelemetryDataSource` but didn't see a way to unwrap it.